make sensor and binary sensor work again!

regevbr opened the pull request:
Both sensors are not working properly now for a number of reasons:
1. The super call in the constructor is not located properly causing runtime issues
2. The unique id was not set up properly causing devices with multiple sensors to only be able to produce a single sensor
3. Sensors with a negative scale caused run-time errors

This PR fixes all of that + fix an issue with the names not being properly set

i.e.
This PR replaces and extends #513 and #514

pabigot reviewed (commented):
I'm wondering whether this and [#497](https://github.com/tuya/tuya-home-assistant/pull/497/commits/b85fbc695f72637ab88d0aedaea60b229846ef66#r721285712)  both do local fixes to a more fundamental problem: that `TuyaHaDevice` should have `homeassistant.helpers.Entity` as a base, which would provide at least some of this.

tetsuo55 commented:
@regevbr looking at the other .py files it seems like this issue is present in many of them.

Could you expand this PR to globally fix this issue?

regevbr commented:
Sure will do

On Tue, Oct 5, 2021, 10:02 tetsuo55 ***@***.***> wrote:

> @regevbr <https://github.com/regevbr> looking at the other .py files it
> seems like this issue is present in many of them.
>
> Could you expand this PR to globally fix this issue?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tuya/tuya-home-assistant/pull/518#issuecomment-934124524>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABMCBHYG7GHLUWHPTSDZ64TUFKPIVANCNFSM5FGL6RJQ>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>

regevbr commented:
@tetsuo55 there was an actual issue just with the alarm which I fixed. For the sake of consistency, I also changed the super call in other classes (currently they don't have an issue)

tetsuo55 commented:
> @tetsuo55 there was an actual issue just with the alarm which I fixed. For the sake of consistency, I also changed the super call in other classes (currently they don't have an issue)

ah i thought the wrong location of the super would be an issue everywhere

regevbr commented:
@tetsuo55 the issue in the alarm was not because of the super but the scenario in which a single device has a couple of sensors, only one of them will be created, I fixed that by adding the sensor name to the unique id and name of the entity
